Use space query instead of explore for delegation page

samuveth opened the pull request:
Fixes https://github.com/snapshot-labs/snapshot/issues/1467

Changes proposed in this pull request:
- Add new space query for delegation page
- Query space on every user input with a .5s debounce

mktcode commented on src/views/Delegate.vue:
```diff
@@ -318,7 +330,7 @@ onMounted(async () => {
           @click="handleSubmit"
           :disabled="!isValid || !$auth.isAuthenticated.value"
-          :loading="loading"
+          :loading="loading || networkQueryLoading"
```
The submit button should open the connect modal when not connected yet. Currently it is just disabled and it took me a few seconds to realize the reason.

mktcode reviewed (commented):
- After adding/removing a delegation, the list doesn't refresh anymore.
- UI flickering:

![ezgif-6-7604ed37ad47](https://user-images.githubusercontent.com/6792578/136982256-394e51d4-8e4b-46cf-936d-56a5bcf46444.gif)

- I'd maybe add a note like: "Can't find your delegations? Make sure you are connected to the correct network. (Delegations need to be made per network.)" Something like that? Or are delegations only meant to be on mainnet and then applied globally?

samuveth commented:
> * After adding/removing a delegation, the list doesn't refresh anymore.

Not sure why this is happening sometimes, it doesn't seem to be related to any of my changes. Maybe the 3 seconds sleep is not enough delay?

> * I'd maybe add a note like: "Can't find your delegations? Make sure you are connected to the correct network. (Delegations need to be made per network.)" Something like that? Or are delegations only meant to be on mainnet and then applied globally?

I need to think about this one more.

samuveth commented:
Fixed the other comments and added some additional features.

- Added a checkbox for delegating to a specific space, I think is makes things more clear for the user. It's switched on by defaults if the url has a key param like `delegation/balancer.eth` 
![Jan-20-2022 17-58-48](https://user-images.githubusercontent.com/51686767/150327866-dc24354c-af7e-4b60-bcc2-5995fe2a7292.gif)
(text edited specified > specific)


- Error messages for address input:
<img width="534" alt="Screen Shot 2022-01-20 at 5 56 10 PM" src="https://user-images.githubusercontent.com/51686767/150327067-6c96c548-8ef8-4d4a-b59d-7bdcd2514da6.png">
<img width="490" alt="Screen Shot 2022-01-20 at 5 55 59 PM" src="https://user-images.githubusercontent.com/51686767/150327082-db966a28-e6c3-4ed0-8083-85ac18de5dc1.png">


- Error message for space input:
<img width="492" alt="Screen Shot 2022-01-20 at 5 56 36 PM" src="https://user-images.githubusercontent.com/51686767/150327167-26eb0dfd-e62a-4e53-8b77-3c6176b163db.png">

mktcode commented:
- "Address must not match your account" only appears when entering the address but not the ENS name. I can't enter "0x27711f9c07230632F2EE1A21a967a9AC4729E520" but I can enter "mktcode.eth".
- If "Limit delegation to a specific space" is enabled, "Confirm" button should be disabled if space field is empty.

mktcode commented on src/components/RegisterENS.vue:
```diff
@@ -5,18 +5,20 @@
 
 const props = defineProps({
-  modelValue: String
+  modelValue: String,
+  validEnsTlds: Array
```
Why this prop?

mktcode replied:
This prop isn't needed, is it?

```suggestion
  modelValue: String
```

mktcode commented on src/views/Delegate.vue:
```diff
@@ -93,12 +112,21 @@ function clearDelegate(id, delegate) {
+    getDelegationsAndDelegatesLoading.value = false;
     delegates.value = delegatesObj.delegations;
     delegators.value = delegatorsObj.delegations;
   }
```
```suggestion
  } else {
    // user logged out
    delegates.value = [];
    delegators.value = [];
  }
```

mktcode commented on src/views/Delegate.vue:
```diff
@@ -55,68 +59,83 @@ const form = ref({
+  modalOpen.value = true;
+}
+
+watch([networkKey, web3Account], ([valN, prevN], [valA, prevA]) => {
```
It's `(newValues, prevValues)` not `(newAndPrevA, newAndPrevB, ...)`

https://v3.vuejs.org/guide/reactivity-computed-watchers.html#watching-multiple-sources

```suggestion
watch([networkKey, web3Account], ([valN, valA], [prevN, prevA]) => {
```

mktcode commented on src/views/Delegate.vue:
```diff
@@ -55,68 +59,83 @@ const form = ref({
+}
+
+watch([networkKey, web3Account], ([valN, prevN], [valA, prevA]) => {
+  if (valN !== prevN) getDelegationsAndDelegates();
+  if (valA?.toLowerCase() !== prevA) getDelegationsAndDelegates();
```
I'd also combine these ifs.

```suggestion
  if (valN !== prevN || valA?.toLowerCase() !== prevA) getDelegationsAndDelegates();
```

I think it's not really possible to trigger both cases at the same time, (at least not with metamask) but technically it would run twice then.

mktcode commented on src/views/Delegate.vue:
```diff
@@ -55,68 +59,83 @@ const form = ref({
 });
 
-async function load() {
+function clearDelegate(id, delegate) {
```
Personally I'd like `revokeDelegation` more. 🙈

mktcode reviewed (changes requested):
See above. :)

samuveth commented:
@mktcode done :)
